File: practica_9/Objetos_Coloreados/PYTHON/main.py
from abc import ABC, abstractmethod
import random
import math

# Interfaz Coloreado
class Coloreado(ABC):
    @abstractmethod
    def comoColorear(self):
        pass

# Clase abstracta Figura
class Figura(ABC):
    def __init__(self, color):
        self.color = color

    def setColor(self, color):
        self.color = color

    def getColor(self):
        return self.color

    @abstractmethod
    def area(self):
        pass

    @abstractmethod
    def perimetro(self):
        pass

    def __str__(self):
        return f"Figura{{color={self.color}}}"

# Clase Cuadrado
class Cuadrado(Figura, Coloreado):
    def __init__(self, color):
        super().__init__(color)
        self.lado = random.randint(1, 100)

    def area(self):
        return self.lado * self.lado

    def perimetro(self):
        return 4 * self.lado

    def comoColorear(self):
        return f"Colorear los cuatro colores de {self.color}"

    def __str__(self):
        return f"Cuadrado{{color={self.color}, lado={self.lado}}}"

# Clase Circulo
class Circulo(Figura, Coloreado):
    def __init__(self, color):
        super().__init__(color)
        self.radio = random.randint(1, 100)

    def area(self):
        return (self.radio * self.radio) * math.pi

    def perimetro(self):
        return 2 * math.pi * self.radio

    def comoColorear(self):
        return f"Colorear el circulo de {self.color}"

    def __str__(self):
        return f"Circulo{{color={self.color}, radio={self.radio}}}"

# Función principal
def main():
    figuritas = []

    c1 = Cuadrado("rojo")
    c2 = Cuadrado("amarillo")
    c3 = Cuadrado("azul")

    cir1 = Circulo("verde")
    cir2 = Circulo("negro")
    cir3 = Circulo("rosa")

    figuritas.append(c1)
    figuritas.append(c2)
    figuritas.append(c3)
    figuritas.append(cir1)
    figuritas.append(cir2)
    figuritas.append(cir3)

    for f in figuritas:
        print(f)
        print(f.area())
        print(f.perimetro())

        if isinstance(f, Cuadrado):
            print(f.comoColorear())
        elif isinstance(f, Circulo):
            print(f.comoColorear())

        print()

# Ejecutar la función principal
if __name__ == "__main__":
    main()

File: practica_9/Objetos_Coloreados/JAVA/Main.java
import java.util.ArrayList;

interface Coloreado {
    String comoColorear();
}

/////////////////////////////////////////////////////////////////////////

abstract class Figura {
    protected String color;
    
    public Figura() {
        
    }
    
    public void setColor(String color) {
        this.color = color;
    }
    
    public String getColor() {
        return color;
    }
    
    public abstract double area();
    public abstract double perimetro();
    
    @Override
    public String toString() {
        return "Figura{color=" + color +"}";
    }
}

/////////////////////////////////////////////////////////////////////////

class Cuadrado extends Figura implements Coloreado {
    private int lado;
    
    public Cuadrado(String color) {
        this.lado = (int) (Math.random() * 100) + 1;
        this.color = color;
    }
    
    public double area() {
        return lado * lado;
    }
    
    public double perimetro() {
        return 4 * lado;
    }
    
    @Override
    public String comoColorear() {
        return "Colorear los cuatro colores de " + color;
    } 
    
    @Override
    public String toString() {
        return "Cuadrado{color=" + color +", lado=" + lado + "}";
    }
}

/////////////////////////////////////////////////////////////////////////

class Circulo extends Figura implements Coloreado {
    private int radio;
    
    public Circulo(String color) {
        this.radio = (int) (Math.random() * 100) + 1;
        this.color = color;
    }
    
    public double area() {
        return (radio * radio) * Math.PI;
    }
    
    public double perimetro() {
        return 2 * Math.PI * radio;
    }    
    
    @Override
    public String comoColorear() {
        return "Colorear el circulo de " + color;
    }   
    
    @Override
    public String toString() {
        return "Circulo{color=" + color +", radio=" + radio + "}";
    }    
}

/////////////////////////////////////////////////////////////////////////

public class Main
{
	public static void main(String[] args) {
	    ArrayList<Figura> figuritas = new ArrayList<>();
	    
		Cuadrado c1 = new Cuadrado("rojo");
		Cuadrado c2 = new Cuadrado("amarillo");
		Cuadrado c3 = new Cuadrado("azul");
		
		Circulo cir1 = new Circulo("verde");
		Circulo cir2 = new Circulo("negro");
		Circulo cir3 = new Circulo("rosa");
		
		figuritas.add(c1);
		figuritas.add(c2);
		figuritas.add(c3);
		figuritas.add(cir1);
		figuritas.add(cir2);
		figuritas.add(cir3);
		
		for (Figura f : figuritas) {
		    System.out.println(f);
		    System.out.println(f.area());
		    System.out.println(f.perimetro());
		    
            if (f instanceof Cuadrado) {
                System.out.println(((Cuadrado) f).comoColorear());
            } else if (f instanceof Circulo) {
                System.out.println(((Circulo) f).comoColorear());
            }
            
		    System.out.println();
		    
		}
	}
}
